### PR TITLE
SNES: Add Turbo File Twin support

### PR DESCRIPTION
**Pull request by NovaSquirrel**
Implements both TFII and STF modes, and the implementation was based on trying to deliberately explore edge cases and figure out the underlying behavior with tests. Licensed games using Turbo File Twin are currently untested.

There's currently a known inaccuracy where, in STF mode, if you start a write command and read the status twice without writing any data, the second status read seems to exit write mode while also misreporting that it's still active. This needs more research but is such a specific edge case that it shouldn't affect any game that isn't deliberately trying to do something weird.

### CODE REVIEW

**Comment by NovaSquirrel** [comment]
Tested it with `RPG Tsukuru - Super Dante` (which uses STF mode). It's able to detect the Turbo File Twin, and I was able to save a file, view a list of files, pick the file I saved, and load it. Derby Stallion II also let me save and load a file. I have not tested to see if the data loaded is correct, but opening the save in a hex editor seems fine?

File: Core/SNES/Input/AsciiTurboFileTwinTf2.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		bool prevStrobe = _strobe;
+		_strobe = (value & 0x01) == 0x01;
+
+		if(prevStrobe && !_strobe) {
+			RefreshStateBuffer();
+
+			if(_unlocked) {
+				//Perform write and increase position
+				uint8_t ioPort = _console->GetInternalRegisters()->GetIoPortOutput();
+				_data[_position / 8] &= ~(1 << (_position % 8));
```
Comment (asiekierka) [inline Core/SNES/Input/AsciiTurboFileTwinTf2.h]: Since this is copy-pasted twice now, maybe a GetBit and SetBit method that operates in byte arrays would make sense in BitUtilities? This would also make the code more readable.

File: Core/SNES/Input/AsciiTurboFileTwinStf.h [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	{
+		_emu->GetBatteryManager()->SaveBattery(".turbofile_stf.sav", _data, AsciiTurboFileTwinStf::FileSize);
+	}
+
+	void RefreshStateBuffer() override
+	{
+		_stateBuffer = 0b011111110111000000000000 | // Controller type $E, and then disambiguated with $FE
+			((uint8_t)(_writeMode & !_didWriteAnything) << 24) |
+			((uint8_t)_readMode << 25);
+		// Bit 26 is 1 if batteries are missing?
```
Comment (asiekierka) [inline Core/SNES/Input/AsciiTurboFileTwinStf.h]: If this is unknown behaviour that needs verification, please mark it with a `TODO` comment.

File: Core/SNES/Input/AsciiTurboFileTwinStf.h [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+		// Bit 27 is 1 if write protect is on
+		// Bits 28-31 are capacity; 0 = 128 KiB
```
Comment (asiekierka) [inline Core/SNES/Input/AsciiTurboFileTwinStf.h:79]: In what units? Could this be derived from FileSize, so that a way to define a dynamic capacity could potentially be supported in the future?

**Comment by asiekierka** [comment]
> There's currently a known inaccuracy where, in STF mode, if you start a write command and read the status twice without writing any data, the second status read seems to exit write mode while also misreporting that it's still active. This needs more research [...]

As above, if you're aware of such inaccuracies, please mark them with a TODO comment for further research.

**Comment by NovaSquirrel** [comment]
> As above, if you're aware of such inaccuracies, please mark them with a TODO comment for further research.

Noted, though in this case I've already pushed a commit that made the write mode flag match the hardware behavior I'm seeing.

**Comment by asiekierka** [comment]
Ah, okay, sorry then!

File: Core/SNES/Input/AsciiTurboFileTwinStf.h (hunk 3 above)
Comment (NovaSquirrel) [reply Core/SNES/Input/AsciiTurboFileTwinStf.h:79]: I referenced this table from fullsnes: https://problemkaputt.de/fullsnes.htm#snesaddonturbofilestfmodefilesystem

nocash's notes are presumably from reverse engineering the code of games that use the Turbo File Twin because we only ever got a 128 KiB model. We could treat the game behavior as authoritative if we do want to eventually support a larger capacity in STF mode, but I wasn't sure about putting data tables in the source code that I can only verify through game behavior. I can confirm I'm getting zeros on actual hardware though.

File: Core/SNES/SnesConsole.cpp [hunk 4]
```diff
@@ -270,6 +270,10 @@ void SnesConsole::SaveBattery()
 	if(_cart) {
 		_cart->SaveBattery();
 	}
+
+	if(_controlManager) {
```
Comment (SourMesen) [inline Core/SNES/SnesConsole.cpp]: It should be possible to move this to Emulator.cpp (and remove this code from both NesConsole & SnesConsole)
This could be done by adding a new SaveBattery function in Emulator, e.g:
```
void Emulator::SaveBattery()
{
	_console->SaveBattery();
	if(_controlManager) {
		_controlManager->SaveBattery();
	}
}
```

and then replacing the 2 calls to `_console->SaveBattery();` in Emulator to just `SaveBattery();`

**Comment by TakuikaNinja** [comment]
Wizardry V - Saika no Chuushin (Japan) (En,Ja)_000
Wizardry V - Saika no Chuushin (Japan) (En,Ja)_001
Wizardry V - Saika no Chuushin (Japan) (En,Ja)_002
Wizardry V - Saika no Chuushin (Japan) (En,Ja)_003

I've been testing the implementation by attempting to import party data from an FC Wizardry game to SFC Wizardry V. (Note: PG = Proving Grounds of the Mad Overlord) Loading the existing data and importing party members seems to work, but not the data save/erase operations (see 4th image)

**Comment by TakuikaNinja** [comment]
Oops, maybe I should use the official English setting.
Wizardry V - Saika no Chuushin (Japan) (En,Ja)_004Wizardry V - Saika no Chuushin (Japan) (En,Ja)_006
Wizardry V - Saika no Chuushin (Japan) (En,Ja)_007Wizardry V - Saika no Chuushin (Japan) (En,Ja)_005

**Comment by TakuikaNinja** [comment]
Here are some test save files for Wizardry V. `Fresh` is the data after attempting (and failing) to save TFII data for the first time, and `PG` is the data from Proving Grounds of the Mad Overlord.
[WizardryV_TFII_Fresh.zip](https://github.com/user-attachments/files/27689102/WizardryV_TFII_Fresh.zip)
[WizardryV_TFII_PG.zip](https://github.com/user-attachments/files/27689105/WizardryV_TFII_PG.zip)

**Comment by NovaSquirrel** [comment]
Thanks, I'll do some tests to try and verify that my logic for when to move onto the next bit is actually correct. If anything is wrong it's probably that.

**Comment by TakuikaNinja** [comment]
Looks like that did the trick - I was able to erase the PG data, save new data (listed as HOM), and load it.
Wizardry V - Saika no Chuushin (Japan) (En,Ja)_008
Wizardry V - Saika no Chuushin (Japan) (En,Ja)_009

**Comment by NovaSquirrel** [comment]
There's some weird behavior that happens when you increment the unlock counter without doing non-strobe $4017 reads in order to actually fully unlock; I've been trying to figure out the logic behind it but I haven't been able to yet. Actual games do this read so they won't be affected, and I don't think this should hold up the merge, but definitely warrants a TODO comment in the code.

File: Core/SNES/Input/AsciiTurboFileTwinStf.h [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+#include "Shared/Interfaces/IBattery.h"
+#include "Utilities/Serializer.h"
+
+class AsciiTurboFileTwinStf : public BaseControlDevice, public IBattery
+{
+private:
+	// Data access
+	static constexpr int FileSize = 128 * 1024;
+	uint32_t _position = 0;
+	uint8_t _data[AsciiTurboFileTwinStf::FileSize] = {};
```
Comment (SourMesen) [inline Core/SNES/Input/AsciiTurboFileTwinStf.h]: Putting large arrays in classes/structs has been a problem at least once in the past (in the old mesen), some compilers seemed to struggle with the TraceLogger class because of it. Though in that case, it was probably a fair bit bigger than this specific array.

I wonder if we should do `new` in the constructor and `delete` in the destructor in this case to avoid any potential issues? 128kb isn't exactly small and everything else larger than a few KBs is usually handled with new/delete atm.
